File: 2020/csharp/Aoc2020.Tests/Day04Tests.cs
using System.Linq;
using System.Collections.Generic;
using Xunit;

namespace Aoc2020.Tests
{
    public class Day04Tests
    {
        [Fact]
        public void TestDataWorks()
        {
            Assert.Equal(2,
                Passport.ParseMany(Utils.ReadInputText("04-test"))
                .Count(passport => passport.AreRequiredFieldsPresent()));

            Assert.Equal(4,
                Passport.ParseMany(Utils.ReadInputText("04-valid"))
                .Count(passport => passport.IsValid()));

            Assert.Equal(0,
                Passport.ParseMany(Utils.ReadInputText("04-invalid"))
                .Count(passport => passport.IsValid()));
        }

        [Fact]
        public void Stars()
        {
            var run = new Day04();
            Assert.Equal("190", run.Result1());
            Assert.Equal("121", run.Result2());
        }
    }
}

File: 2015/fsharp/Aoc2015.Tests/Day04.Tests.fs
module Day04.Tests

open Xunit
open Utils
open Day04

[<Theory>]
[<InlineData(609043, "abcdef")>]
[<InlineData(1048970, "pqrstuv")>]
let ``Day04 Part 1`` expected secret =
    Assert.Equal(expected, secret |> findLowestNumberWithPrefix "00000")

[<Fact>]
let ``Day04 Stars`` () =
    try
        Assert.Equal("254575", job1 ())
        Assert.Equal("1038736", job2 ())
    with :? System.NotImplementedException ->
        ()
